### PR TITLE
Do lockfile validation before applying it

### PR DESCRIPTION
**Pull request by ChengTian**
parent #1220 

Sample `frameworkDependencies` property in `project.lock.json`:
```json
  "frameworkDependencies": {
    "": [
      "EntityFramework.SqlServer >= 7.0.0-*",
      "EntityFramework.InMemory >= 7.0.0-*",
      "Kestrel >= 1.0.0-*",
      "Microsoft.AspNet.Diagnostics >= 1.0.0-*",
      "Microsoft.AspNet.Diagnostics.Entity >= 7.0.0-*",
      "Microsoft.AspNet.Identity.EntityFramework >= 3.0.0-*",
      "Microsoft.AspNet.Mvc >= 6.0.0-*",
      "Microsoft.AspNet.Security.Cookies >= 1.0.0-*",
      "Microsoft.AspNet.Security.Facebook >= 1.0.0-*",
      "Microsoft.AspNet.Security.Google >= 1.0.0-*",
      "Microsoft.AspNet.Security.MicrosoftAccount >= 1.0.0-*",
      "Microsoft.AspNet.Security.OpenIdConnect >= 1.0.0-*",
      "Microsoft.AspNet.Security.Twitter >= 1.0.0-*",
      "Microsoft.AspNet.Server.IIS >= 1.0.0-*",
      "Microsoft.AspNet.Server.WebListener >= 1.0.0-*",
      "Microsoft.AspNet.Session >= 1.0.0-*",
      "Microsoft.AspNet.SignalR.Server >= 3.0.0-*",
      "Microsoft.AspNet.StaticFiles >= 1.0.0-*",
      "Microsoft.Framework.Cache.Distributed >= 1.0.0-*",
      "Microsoft.Framework.CodeGenerators.Mvc >= 1.0.0-*",
      "Microsoft.Framework.ConfigurationModel.Json >= 1.0.0-*",
      "Microsoft.Framework.Logging.Console >= 1.0.0-*"
    ],
    "Asp.Net,Version=v5.0": [
      "framework/System.Collections ",
      "framework/System.IO ",
      "framework/System.IO.Compression ",
      "framework/System.Net.Http ",
      "framework/System.Runtime ",
      "framework/System.Runtime.Serialization ",
      "framework/System.Security ",
      "framework/System.Xml ",
      "framework/System.Xml.Linq "
    ],
    "Asp.NetCore,Version=v5.0": [
      "Microsoft.Net.Http.Client >= 1.0.0-*",
      "System.IO.FileSystem.Primitives >= 4.0.0-beta-*",
      "System.Linq.Parallel >= 4.0.0-beta-*",
      "System.Reflection.TypeExtensions >= 4.0.0-beta-*",
      "System.Runtime.Serialization.Primitives >= 4.0.0-beta-*",
      "System.Security.Cryptography.Hashing.Algorithms >= 4.0.0-beta-*",
      "System.Text.RegularExpressions >= 4.0.10-beta-*"
    ]
  },
```

### CODE REVIEW

File: src/Microsoft.Framework.Runtime/ApplicationHostContext.cs [hunk 1]
```diff
@@ -99,6 +110,55 @@ public class ApplicationHostContext
             _serviceProvider.Add(typeof(ICompilerOptionsProvider), compilerOptionsProvider);
         }
 
+        private bool IsValidLockFile(LockFile lockFile)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/ApplicationHostContext.cs]: This needs to be super fast and it doesn't look fast right now.

**Comment by davidfowl** [comment]
The name frameworkDependencies isn't a good one. We need something like "project.json"

**Comment by ChengTian** [comment]
Optimized. It's so fast that dotTrace cannot capture `IsValidLockFile` :smile: 
afteroptimization

File: src/Microsoft.Framework.Runtime/DependencyManagement/LockFile.cs [hunk 2]
```diff
@@ -9,6 +9,8 @@ namespace Microsoft.Framework.Runtime.DependencyManagement
     public class LockFile
     {
         public bool Islocked { get; set; }
+        public IList<Tuple<string, IEnumerable<string>>> ProjectFileDependencies { get; set; } =
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/DependencyManagement/LockFile.cs]: Create a type for this, this is ugly and it needs to be back ported to NuGet.ProjectModel /cc @anurse

Comment (ChengTian) [inline src/Microsoft.Framework.Runtime/DependencyManagement/LockFile.cs]: Created new type for it.

**Comment by ChengTian** [comment]
Will port to NuGet repo after this one gets sign-off

**Comment by ChengTian** [comment]
Fixed class naming

File: src/Microsoft.Framework.Runtime/ApplicationHostContext.cs [hunk 3]
```diff
@@ -42,22 +44,39 @@ public class ApplicationHostContext
             ProjectDepencyProvider = new ProjectReferenceDependencyProvider(ProjectResolver);
             var unresolvedDependencyProvider = new UnresolvedDependencyProvider();
 
+            var projectName = PathUtility.GetDirectoryName(ProjectDirectory);
+
+            Project project;
+            if (ProjectResolver.TryResolveProject(projectName, out project))
+            {
+                Project = project;
+            }
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/ApplicationHostContext.cs]: Check for null project, make sure there;s no null ref.

File: src/Microsoft.Framework.Runtime/ApplicationHostContext.cs [hunk 4]
```diff
@@ -42,22 +44,39 @@ public class ApplicationHostContext
...
+
             var projectLockJsonPath = Path.Combine(ProjectDirectory, LockFileFormat.LockFileName);
-            if (File.Exists(projectLockJsonPath))
+            var lockFileExists = File.Exists(projectLockJsonPath);
+            var validLockFile = false;
+
+            if (lockFileExists)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/ApplicationHostContext.cs]: Add some comments.

File: src/Microsoft.Framework.Runtime/ApplicationHostContext.cs [hunk 5]
```diff
@@ -107,6 +118,45 @@ public class ApplicationHostContext
             _serviceProvider.Add(typeof(ICompilerOptionsProvider), compilerOptionsProvider);
         }
 
+        private bool IsValidLockFile(LockFile lockFile)
+        {
+            // The lock file should contain dependencies for each framework plus dependencies shared by all frameworks
+            if (lockFile.ProjectFileDependencies.Count != Project.GetTargetFrameworks().Count() + 1)
```
Comment (davidfowl) [inline src/Microsoft.Framework.Runtime/ApplicationHostContext.cs]: `Project.GetTargetFrameworks()` gets called twice and we do `ToDictionary` for a pretty small number of

**Comment by ChengTian** [comment]
Made code review changes

**Comment by ChengTian** [comment]
Got personal sign-off from @davidfowl , pushing to dev

**Pull request merged**